crates/store/src/state.rs

```diff
@@ -494,6 +496,145 @@ impl State {
         Ok(NoteAuthenticationInfo { block_proofs, note_proofs })
     }
 
+    /// Fetches the inputs for a transaction batch from the database.
+    ///
+    /// ## Inputs
+    ///
+    /// The function takes as input:
+    /// - The tx reference blocks are the set of blocks referenced by transactions in the batch.
+    /// - The unauthenticated note ids are the set of IDs of unauthenticated notes consumed by all
+    ///   transactions in the batch. For these notes, we attempt to find note inclusion proofs. Not
+    ///   all notes will exist in the DB necessarily, as some notes can be created and consumed
+    ///   within the same batch.
+    ///
+    /// ## Outputs
+    ///
+    /// The function will return:
+    /// - A block inclusion proof for all tx reference blocks and for all blocks which are
+    ///   referenced by a note inclusion proof.
+    /// - Note inclusion proofs for all notes that were found in the DB.
+    /// - The block header that the batch should reference, i.e. the latest known block.
+    pub async fn get_batch_inputs(
+        &self,
+        tx_reference_blocks: BTreeSet<BlockNumber>,
+        unauthenticated_note_ids: BTreeSet<NoteId>,
+    ) -> Result<BatchInputs, GetBatchInputsError> {
+        if tx_reference_blocks.is_empty() {
+            return Err(GetBatchInputsError::TransactionBlockReferencesEmpty);
+        }
+
+        // First we grab note inclusion proofs for the known notes. These proofs only
+        // prove that the note was included in a given block. We then also need to prove that
+        // each of those blocks is included in the chain.
+        let note_proofs = self
+            .db
+            .select_note_inclusion_proofs(unauthenticated_note_ids)
+            .await
+            .map_err(GetBatchInputsError::SelectNoteInclusionProofError)?;
+
+        // The set of blocks that the notes are included in.
+        let note_blocks = note_proofs.values().map(|proof| proof.location().block_num());
+
+        // Collect all blocks we need to query without duplicates, which is:
+        // - all blocks for which we need to prove note inclusion.
+        // - all blocks referenced by transactions in the batch.
+        let mut blocks = tx_reference_blocks;
+        blocks.extend(note_blocks);
+
+        // Grab the block merkle paths from the inner state.
+        //
+        // NOTE: Scoped block to automatically drop the mutex guard asap.
+        //
+        // We also avoid accessing the db in the block as this would delay
+        // dropping the guard.
+        let (batch_reference_block, partial_mmr) = {
+            let state = self.inner.read().await;
+            let latest_block_num = state.latest_block_num();
+
+            let highest_block_num =
+                *blocks.last().expect("we should have checked for empty block references");
```

Comment:
Question: why do we assume that the last element in `blocks` will be the highest block? I think this is because we know that `.db.select_note_inclusion_proofs()` returns note proofs ordered by block number? But what if one of the blocks in `tx_reference_blocks` list has a higher value than any of the note blocks?

Maybe instead of `blocks.last()` this should be `blocks.iter().max()`?

Reply:
`blocks` is a `BTreeSet<BlockNumber>`, so it is ordered.
And a couple lines earlier, we define this set to consist of all blocks referenced by transactions and all blocks referenced by note blocks, so `blocks.last()` will contain the highest block number that is referenced by anything in the batch.
`BTreeSet::last` is documented to return the maximum element of the set: https://doc.rust-lang.org/std/collections/struct.BTreeSet.html#method.last

Reply:
Ah! For some reason I thought it was a `vec`.